Specific error output

m-pitera opened the pull request:
implemented the story that asked for more specific response to users that have incorrect input for their guesses.

There was some refactoring and implementation of feedback in most recent commit

m-pitera committed:
WIP specific error output

m-pitera committed:
wrote validation tests and validate skeleton

m-pitera committed:
implemented logic to output errors through validate

m-pitera committed:
filling in helper methods of is_valid?

m-pitera committed:
wrapped up implementation of validation

fixed some syntax within Validate module. Called the module within user. Cleaned up some logic within user. Also changed error output formatting

m-pitera committed:
turned Output and Colors into a modules as well as correctly implemented GameStatus as a module

ssunday commented on lib/output.rb:
```diff
@@ -2,106 +2,124 @@
+  def type_ok
+    print ("Type 'ok' to continue: ")
+    response = gets.chomp
+    if response.downcase.eql? 'ok'
```
unless response.downcase == 'ok'
 type_ok
end

m-pitera replied:
Right, guard cases; thx for pointing that out.

Is there any reason you prefer == over .eql? or is it simply convention?

ryanzverner replied:
They're _almost_ always interchangeable, and for your comparison here, using `==` would be more conventional.

ssunday commented on lib/validate.rb:
```diff
@@ -0,0 +1,58 @@
+    private
+
+  def length_valid(the_input)
+    return true if the_input.length.eql? 4
```
the_input.length == 4 for the entire method is cleaner

ssunday commented on lib/validate.rb:
```diff
@@ -0,0 +1,58 @@
+  end
+
+  def type_valid(the_input)
+    return true if the_input.all? { |val| val.is_a? Integer}
```
same thing...all of these can be reduced to their checks

ryanzverner commented:
@m-pitera Can you create a README that explains how someone can play the game on the command line?

soft-circles commented on lib/game_status.rb:
```diff
@@ -2,32 +2,30 @@
+    puts(Colors.cyan_b('Thanks for playing!'))
+    exit(0)
   end
 end
```
In a `game_status` class I would expect there to be public methods like `game_over?`. When thinking about game states, try to think about the questions you're interesting in asking. "Is the game over?", "How many guesses are left?" etc. Let these methods answer these questions really simply, usually with just a boolean. Let other methods rely on these game-state methods to handle branching behavior like continuing the game or ending the game.

soft-circles commented on lib/user.rb:
```diff
@@ -1,8 +1,9 @@
 # User class knows too much, i made a mess and now refactoring
 class User
+  include Validate, Output, GameStatus
   attr_reader :guess, :guesses_left, :first_time
```
User has a lot of dependencies... Why is `User` relying on Output?

soft-circles commented on lib/user.rb:
```diff
@@ -63,17 +57,6 @@ def next_command(previous)
-
   def is_quit?(user_input)
     if (user_input.downcase.eql? 'quit')
       # quit_plz
```
So it looks like `User` is still doing way too much. Remember to ask yourself the question "What is the responsibility of this class?" and answer in a simple sentence. If you use the word "and" anywhere in that sentence, that should set off a red flag.

soft-circles replied:
Also, where are your tests for `User`? ❓

soft-circles commented on lib/validate.rb:
```diff
@@ -0,0 +1,58 @@
+          print_type_error
+        elsif error.eql? 'context'
+          print_context_error
+        end
```
Let's break off the printing of each error into a seperate method. I like the chaining you're doing here but I think you can separate the handling of recording errors and printing them out to the console.

ryanzverner commented on lib/colors.rb:
```diff
@@ -1,24 +1,22 @@
+  def cyan(text); colorize(text, 36); end
+  def cyan_b(text); colorize(text, 96); end
+  def white(text); colorize(text, 37); end
+  def white_b(text); colorize(text, 97); end
```
These methods are a bit confusing. They're named as if they simply return a color, perhaps a hex code. But in reality they take a parameter that's some text and then return that text in a particular color. Think about other ways to achieve this.

m-pitera replied:
I implemented a new structure within the commit marked as "changed Colors' architecture", what do you think?

ryanzverner commented on lib/secret_code.rb:
```diff
@@ -1,17 +1,18 @@
     randomness = Random.new
     return (1..4).map { |_| randomness.rand(1..6) }
-    Output.print_code_generated
+    print_code_generated
```
This line will never execute.

m-pitera replied:
good call x_x

zdavy commented on lib/output.rb:
```diff
@@ -2,106 +2,124 @@
-      system 'clear'
-    end
+module Output
+  include Colors
```
A really cool thing about modules is that they can share behavior with a given object, but it's important to be diligent in it's use to avoid violating the [Liskov Substitution Principle](https://www.tomdalling.com/blog/software-design/solid-class-design-the-liskov-substitution-principle/).

I like the *is/behaves like/has* smell test, it helps decide how to include external behavior into a class. The most coherent phrasing would (respectively) indicate either to use inheritance, mix-ins, or dependency injection. If none of the sentences make sense, I would create a higher level abstraction to coordinate between the namespaces.

So in this situation, our three indicators would be
* output _is_ colors
* output _behaves like_ colors
* output _has_ colors

While _output behaves like colors_ sounds OK, I think that _output has colors_ makes the most sense in this given situation, so IMO both `Output` and `Colors` might be best suited as classes, with `Colors` as a dependency of the `Output`.

Here's some other semi-related topics to look into:
[Composition over Inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance)
[Modules](http://ruby-doc.com/docs/ProgrammingRuby/html/tut_modules.html)
[Why Mixins are Broken](https://reactjs.org/blog/2016/07/13/mixins-considered-harmful.html#why-mixins-are-broken)

That last shared article talks specifically about reactjs, but the concepts under the "Why Mixins are Broken" header apply to ruby mixins just as well.

Read some of that over, and let me know what you think.

m-pitera replied:
Ok so from the articles you linked and some other stuff I read this is what I understand: Mix-ins are defining methods to be used over potentially multiple classes, but Inheritance is more so laying out an empty or default method(s) that can be overridden over multiple classes. Right?

So in this case you recommend to use neither of those, but instead dependency injection where I would simply create a Colors object within Output and have it run it's methods when it is necessary. Would that be the correct way of implementing dependency injection in this case?

m-pitera committed:
wrote the readme

m-pitera committed:
reverted some changes so game runs

the changes 2 commits ago where Colors became a module caused errors in Output. Output being a module also can't store instance variables, which was used in order for the Code to be printed at the end. It was terrible implementation in the first place, but now it caused problems. Also forgot to require GameStatus into User at some point. Overall this is all a huge mess, but i need to implement feedback first b4 refactoring again so i can merge this branch

m-pitera committed:
shortened conditionals and replaced if with guard clauses

m-pitera committed:
commented out 'type' section of valid

user input will always be a string . . .

m-pitera committed:
Output (module -> class)

implementing 'has a' relationship rather than 'behaves like'/'is a'

m-pitera committed:
changed Colors' architecture

cleaned up Colors class with a hash rather than 20 methods. Had to go through and change every instance where a Color method was called, but it's good now, i hope.

m-pitera committed:
extracted part of is_valid into a private method

is_valid was both accessing the errors and printing them so I extracted some code into a private method that accesses which methods to print and notifies Output

ssunday commented on README.md:
```diff
@@ -1 +1,5 @@
 # Mastermind
+
+  In order to run the game simply clone or download the repo and you can run it though "ruby main.rb" when within the repo's root directory
```
`ruby main.rb`

Usually readme's are have a setup section and the first paragraph is like a description or abstract.

Maybe include how to run specs?

m-pitera replied:
Aight, thx

ssunday commented on lib/validate.rb:
```diff
@@ -0,0 +1,51 @@
+      errors << 'length'
+    end
+
+    # if !type_valid(the_input)
```
Is this going to be put back in or is it plain junk?

m-pitera replied:
i think it's plain junk, it serves no purpose. I'll delete it then

m-pitera committed:
created GameFunctions class

GameStatus now only handles determining the 'status' of the game and whether it should be over. Sends msg to GameFunctions calling for the game to end if the situation is neccessary

m-pitera committed:
revised readme

also deleted commented out code in validate

ssunday commented on README.md:
```diff
@@ -1,5 +1,23 @@
+
+  You can run specs by running this command while in the root directory:
+
+  `rspec`
```
😁
